# **Dados:**

 **Aluno(a):** `Lucas Ariel Alves de Carvalho`

 **Matrícula:** `121210801`


# Objetivo:

Nesse laboratório, você extrairá um romance de um arquivo HTML retirado do site Projeto Gutenberg (que contém um grande corpus de livros) usando o pacote de requisições do Python. Em seguida, você normalizará e tokenizará o texto. Você também irá analisar a distribuição de palavras. Ao final você deverá simular a execução do algoritmo BPE (Byte-Pair Encoding) e calcular a distância de edição entre duas palavras.

# **Bibliotecas**

In [34]:
import requests
from bs4 import BeautifulSoup
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer, WordNetLemmatizer
from nltk.tokenize import RegexpTokenizer

nltk.download('stopwords')
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

# **Obtenção dos dados**

Para a atividade desse laborátorio, iremos utilizar dados textuais oriundos do **[Project Gutenberg](https://www.gutenberg.org/)**. O texto escolhido é a obra Moby Dick.

In [2]:
#Endereço do texto
r = requests.get("https://www.gutenberg.org/files/2701/2701-h/2701-h.htm")

r.encoding = 'utf-8'

html = r.text

print(html[0:2000])

<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
<title>Moby Dick; or The Whale</title>

<style>

    body {margin-left:10%; margin-right:10%; text-align:justify }
    p { text-indent: 1em; margin-top: .25em; margin-bottom: .25em; }
    H1,H2,H3,H4,H5,H6 { text-align: center; margin-left: 15%; margin-right: 15%; }
    hr  { width: 50%; text-align: center;}
    blockquote {font-size: 100%; margin-left: 0%; margin-right: 0%;}
    .mynote    {background-color: #DDE; color: #000; padding: .5em; margin-left: 10%; margin-right: 10%; font-family: sans-serif; font-size: 95%;}
    .toc       { margin-left: 10%; margin-bottom: .75em;}
    pre        { font-family: times new roman; font-size: 100%; margin-left: 10%;}

    table      {margin-left: 10%;}

a:link {color:blue;
		text-decoration:none}
link {color:blue;
		text-decoration:none}
a:visited {color:blue;
		text-decoration:none}
a:hover {color:red}

</style>
  </head>
  <body>
<div>*** START OF THE PROJECT GUTENBERG EBOOK 270

In [4]:
# Exibir o livro
soup = BeautifulSoup(html, 'html.parser')

text = soup.get_text()

print(text[32000:34000])

h her surf.
      Right and left, the streets take you waterward. Its extreme downtown is
      the battery, where that noble mole is washed by waves, and cooled by
      breezes, which a few hours previous were out of sight of land. Look at the
      crowds of water-gazers there.
    

      Circumambulate the city of a dreamy Sabbath afternoon. Go from Corlears
      Hook to Coenties Slip, and from thence, by Whitehall, northward. What do
      you see?—Posted like silent sentinels all around the town, stand
      thousands upon thousands of mortal men fixed in ocean reveries. Some
      leaning against the spiles; some seated upon the pier-heads; some looking
      over the bulwarks of ships from China; some high aloft in the rigging, as
      if striving to get a still better seaward peep. But these are all
      landsmen; of week days pent up in lath and plaster—tied to counters,
      nailed to benches, clinched to desks. How then is this? Are the green
      fields gone? What do

# **Pré-Processamento**
O resultado de cada uma das etapas a seguir **deve ser utilizada como entrada para as etapas posteriores!!!**

In [60]:
#Inicialmente normalize o texto, tornando-o completamente minúsculo. Você pode fazer isso usando a função lower() da classe string.
def normalizar(texto):
    return texto.lower()

texto_normalizado = normalizar(text)
print(texto_normalizado[0:50])





moby dick; or the whale



*** start of the pr


In [9]:
# Tokenize o texto considerando apenas caracteres alpha-numéricos. Você pode fazer isso com expressões regulares e a biblioteca nltk.tokenize.RegexpTokenizer() passando seu padrão ou re.find()
def tokenizar(texto):
   tokenizer = RegexpTokenizer(r'\w+')
   return tokenizer.tokenize(texto)


tokens = tokenizar(texto_normalizado)

print(f"Número de tokens: {len(tokens)}")
print("Primeiros 20 tokens:", tokens[:20])

Número de tokens: 219426
Primeiros 20 tokens: ['moby', 'dick', 'or', 'the', 'whale', 'start', 'of', 'the', 'project', 'gutenberg', 'ebook', '2701', 'moby', 'dick', 'or', 'the', 'whale', 'by', 'herman', 'melville']


In [13]:
# Implemente o código para remover tokens que possuem tamanho maior que 15 e menor que 2. Você pode fazer isso usando a função len() para verificar o tamanho das palavras
def remove_palavras_pequenas_grandes(tokens):
    return [token for token in tokens if 2 <= len(token) <= 15]


process_1 = remove_palavras_pequenas_grandes(tokens)
print(f"Número de tokens: {len(process_1)}")
print("Primeiros 20 tokens:", process_1[:20])

Número de tokens: 210173
Primeiros 20 tokens: ['moby', 'dick', 'or', 'the', 'whale', 'start', 'of', 'the', 'project', 'gutenberg', 'ebook', '2701', 'moby', 'dick', 'or', 'the', 'whale', 'by', 'herman', 'melville']


In [14]:
# Implemente esta função para ela retornar uma lista de tokens sem stopwords. Você pode fazer isso usando a função stopwords.words('english') para encontrar todas as stopwords
def remove_stopwords(tokens):
  stop_words = set(stopwords.words('english'))
  return [token for token in tokens if token not in stop_words]


process_2 = remove_stopwords(process_1)
print(f"Número de tokens: {len(process_2)}")
print("Primeiros 20 tokens:", process_2[:20])

Número de tokens: 111191
Primeiros 20 tokens: ['moby', 'dick', 'whale', 'start', 'project', 'gutenberg', 'ebook', '2701', 'moby', 'dick', 'whale', 'herman', 'melville', 'contents', 'etymology', 'extracts', 'supplied', 'sub', 'sub', 'librarian']


In [15]:
# Implemente esta função para ela retornar uma lista com os tokens stemmizados. Você pode fazer isso usando a função stem() da classe PorterStemmer e passando para ela a palavra desejada
def stemming(tokens):
  stemmer = PorterStemmer()
  return [stemmer.stem(token) for token in tokens]


process_3 = stemming(process_2)
print(f"Número de tokens: {len(process_3)}")
print("Primeiros 20 tokens:", process_3[:20])

Número de tokens: 111191
Primeiros 20 tokens: ['mobi', 'dick', 'whale', 'start', 'project', 'gutenberg', 'ebook', '2701', 'mobi', 'dick', 'whale', 'herman', 'melvil', 'content', 'etymolog', 'extract', 'suppli', 'sub', 'sub', 'librarian']


In [16]:
# Implemente esta função para ela retornar uma lista com os tokens lematizados. Você pode fazer isso usando a função lemmatize() da classe WordNetLemmatizer e passando para ela a palavra desejada
def lemmatize(texto):
  lemmatizer = WordNetLemmatizer()
  return [lemmatizer.lemmatize(token) for token in texto]


process_4 = lemmatize(process_3)
print(f"Número de tokens: {len(process_4)}")
print("Primeiros 20 tokens:", process_4[:20])

Número de tokens: 111191
Primeiros 20 tokens: ['mobi', 'dick', 'whale', 'start', 'project', 'gutenberg', 'ebook', '2701', 'mobi', 'dick', 'whale', 'herman', 'melvil', 'content', 'etymolog', 'extract', 'suppli', 'sub', 'sub', 'librarian']


## **Frequência dos tokens**
Mostre a distribuição das palavras para cada uma das etapas de pré-processamento realizadas. Você pode fazer isso utilizando a função nltk.FreqDist() passando sua lista de palavras

In [55]:
#Frequência das palavras depois de normalizadas e das pequenas serem removidas
from nltk import FreqDist
frequencia = FreqDist(tokens)
frequencia.most_common(20)


[('the', 14538),
 ('of', 6626),
 ('and', 6447),
 ('a', 4747),
 ('to', 4627),
 ('in', 4184),
 ('that', 3085),
 ('his', 2532),
 ('it', 2522),
 ('i', 2127),
 ('he', 1897),
 ('but', 1818),
 ('s', 1812),
 ('as', 1742),
 ('is', 1725),
 ('with', 1723),
 ('was', 1645),
 ('for', 1618),
 ('all', 1527),
 ('this', 1396)]

In [47]:
#Frequência das palavras tokenizadas
from nltk import FreqDist
frequencia2 = FreqDist(process_2)
frequencia2.most_common(20)


[('whale', 1240),
 ('one', 923),
 ('like', 647),
 ('upon', 566),
 ('man', 527),
 ('ship', 519),
 ('ahab', 517),
 ('ye', 473),
 ('sea', 455),
 ('old', 450),
 ('would', 432),
 ('though', 384),
 ('head', 348),
 ('yet', 345),
 ('boat', 337),
 ('time', 334),
 ('long', 333),
 ('captain', 329),
 ('still', 312),
 ('chapter', 308)]

In [48]:
#Frequência das palavras tokenizadas e com stemming
from nltk import FreqDist
frequencia3 = FreqDist(process_3)
frequencia3.most_common(20)

[('whale', 1646),
 ('one', 943),
 ('like', 661),
 ('ship', 625),
 ('upon', 566),
 ('ye', 548),
 ('sea', 542),
 ('man', 541),
 ('ahab', 518),
 ('boat', 484),
 ('head', 469),
 ('seem', 463),
 ('old', 450),
 ('time', 449),
 ('would', 432),
 ('though', 384),
 ('hand', 368),
 ('look', 366),
 ('captain', 353),
 ('yet', 345)]

In [59]:
#Frequência das palavras tokenizadas, com stemming e lemming
from nltk import FreqDist
frequencia4 = FreqDist(process_4)
frequencia4.most_common(20)

[('whale', 1646),
 ('one', 943),
 ('like', 661),
 ('ship', 625),
 ('upon', 566),
 ('ye', 548),
 ('sea', 542),
 ('man', 541),
 ('ahab', 518),
 ('boat', 484),
 ('head', 469),
 ('seem', 463),
 ('old', 450),
 ('time', 449),
 ('would', 432),
 ('though', 384),
 ('hand', 368),
 ('look', 366),
 ('captain', 353),
 ('yet', 345)]

# Perguntas

1. **Compare o texto original com o texto após a normalização. Quais diferenças você percebe nas palavras e na estrutura geral?**
- Resposta: `A normalização converte todo o texto para minúsculas, eliminando diferenças entre maiúsculas e minúsculas que podem representar a mesma palavra. Isso uniformiza o texto mas remove nomes próprios que iniciam com letra maiúscula.`

2. **O que muda na quantidade de tokens ao aplicar a função remove_palavras_pequenas_grandes()? Informe abaixo o total de tokens antes e depois da aplicação da função.**
- Resposta: `Antes: 219426 / Depois: 210173. A função remove palavras muito curtas (como artigos e preposições de uma letra) e muito longas (que podem ser erros ou palavras pouco comuns), tem que estar >=2 e <= 15.`

3. **Existe diferença entre o pré-processamento antes e depois de aplicar lematização e stemming?**
- Resposta: `Sim, o stemming reduz palavras para sua raiz, podendo criar formas não existentes no dicionário, enquanto a lematização transforma palavras para sua forma no dicionario, mantendo palavras válidas.`

# **Algoritmo Byte-Pair Encoding:**

Considerando as três palavras mais frequentes resultantes da lemmatização (process_4), execute, à mão, o algoritmo BPE para k = 3 e insira abaixo o vocabulário e regras obtidas:

Palavras mais frequentes: "whale" (1646), "one" (943), "like" (661)

*   vocabulário obtido: `'w', 'h', 'a', 'l', 'e', 'o', 'n', 'i', 'k', 'wh', 'le', 'ale'`
*   regras obtidas: `
                      + e → le (2307 ocorrências)
                      w + h → wh (1646 ocorrências)
                      a + le → ale (1646 ocorrências)`













# **Distância de Edição**
Calcule à mão a distância de edição entre as palavras 'while' e 'like' supondo que inserções ou deleções tem custo 1 e uma substituição tem custo 2 quando os caracteres são diferentes e custo 0 quando são iguais.


Insira o valor obtido aqui: `Total: 1 + 2 + 0 + 2 + 0 = 5`

Motivo: Remover 'w' (Custo 1) + Substituir 'h' por 'l' (custo 2) + Manter 'i' (custo 0) + Substituir 'l' por 'k' (custo 2) + Manter 'e' (custo 0)